File: SampleApps/ObjC/AdColonyBasic/AdColonyBasic/BannersViewController.m
//
//  BannersViewController.m
//  AdColonyBasic
//
//  Copyright © 2019 AdColony. All rights reserved.
//

#import "BannersViewController.h"
#import "Settings.h"

#import <AdColony/AdColony.h>


#pragma mark - BannersViewController Interface

@interface BannersViewController () <AdColonyAdViewDelegate>

@property (nonatomic, weak) AdColonyAdView *ad;

@property (nonatomic, weak) IBOutlet UIView *bannerPlacement;
@property (nonatomic, weak) IBOutlet UIActivityIndicatorView *spinner;
@property (nonatomic, weak) IBOutlet UILabel *loadingLabel;
@property (nonatomic, weak) IBOutlet UIButton *loadButton;

@end

@implementation BannersViewController

#pragma mark - UIViewController Overrides

- (void)viewDidLoad {
    [super viewDidLoad];
    [self setLoadingState];
}

- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
    [self requestBanner];
}

- (void)viewDidDisappear:(BOOL)animated {
    [super viewDidDisappear:animated];
    [self clearBanner];
}

#pragma mark - AdColony

- (void)requestBanner {
    [self setLoadingState];
    [AdColony requestAdViewInZone:kAdColonyBannerZoneID withSize:kAdColonyAdSizeBanner viewController:self andDelegate:self];
}

- (void)clearBanner {
    if (self.ad) {
        [self.ad destroy];
        self.ad = nil;
    }
}

- (void)adColonyAdViewDidLoad:(AdColonyAdView *)adView {
    [self clearBanner];
    self.ad = adView;
    CGSize placementSize = self.bannerPlacement.frame.size;
    adView.frame = CGRectMake(0, 0, placementSize.width, placementSize.height);
    [self.bannerPlacement addSubview:adView];
    [self setReadyState];
}

- (void)adColonyAdViewDidFailToLoad:(AdColonyAdRequestError *)error {
    NSLog(@"SAMPLE_APP: Banner request failed with error: %@ and suggestion: %@", [error localizedDescription], [error localizedRecoverySuggestion]);
    [self setReadyState];
}

#pragma mark - UI

- (void)setReadyState {
    self.loadingLabel.hidden = YES;
    [self.spinner stopAnimating];
    self.loadButton.hidden = NO;
    [UIView animateWithDuration:1.0 animations:^{ self.loadButton.alpha = 1.0; } completion:nil];
}

- (void)setLoadingState {
    self.loadButton.hidden = YES;
    self.loadButton.alpha = 0.0;
    self.loadingLabel.hidden = NO;
    [self.spinner startAnimating];
}

#pragma mark - Action Handlers

- (IBAction)loadBanner:(id)sender {
    [self requestBanner];
}

- (IBAction)goBack:(id)sender {
    [self dismissViewControllerAnimated:YES completion:nil];
}

@end

File: SampleApps/Swift/AdColonyBasic/AdColonyBasic/BannersViewController.swift
//
//  BannersViewController.swift
//  AdColonyBasic
//
//  Copyright © 2019 AdColony. All rights reserved.
//

import UIKit

class BannersViewController: UIViewController, AdColonyAdViewDelegate {

    @IBOutlet weak var loadButton: UIButton!
    @IBOutlet weak var spinner: UIActivityIndicatorView!
    @IBOutlet weak var loadingLabel: UILabel!
    @IBOutlet weak var bannerPlacement: UIView!
    
    weak var ad: AdColonyAdView? = nil
    
    //=============================================
    // MARK:- UIViewController Overrides
    //=============================================
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.setLoadingState()
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        self.requestBanner()
    }
    
    override func viewDidDisappear(_ animated: Bool) {
        super.viewDidDisappear(animated)
        self.clearBanner()
    }
    
    //=============================================
    // MARK:- AdColony
    //=============================================
    
    func requestBanner() {
        self.setLoadingState()
        AdColony.requestAdView(inZone: Constants.adColonyBannerZoneID, with: kAdColonyAdSizeBanner, viewController: self, andDelegate: self)
    }
    
    func adColonyAdViewDidLoad(_ adView: AdColonyAdView) {
        self.clearBanner()
        self.ad = adView
        let placementSize = self.bannerPlacement.frame.size;
        adView.frame = CGRect(x: 0, y: 0, width: placementSize.width, height: placementSize.height)
        self.bannerPlacement.addSubview(adView)
        self.setReadyState()
    }
    
    func adColonyAdViewDidFail(toLoad error: AdColonyAdRequestError) {
        print("SAMPLE_APP: Banner request failed with error: \(error.localizedDescription) and suggestion: \(error.localizedRecoverySuggestion!)")
        self.setReadyState()
    }
    
    func clearBanner() {
        if let adView = self.ad {
            adView.destroy()
            self.ad = nil
        }
    }
    
    //=============================================
    // MARK:- UI
    //=============================================
    
    func setLoadingState() {
        self.spinner.startAnimating()
        self.loadingLabel.isHidden = false
        self.loadButton.alpha = 0
        self.loadButton.isHidden = true
    }
    
    func setReadyState() {
        self.spinner.stopAnimating()
        self.loadingLabel.isHidden = true
        self.loadButton.isHidden = false
        UIView.animate(withDuration: 1) {
            self.loadButton.alpha = 1
        }
    }
    
    @IBAction func loadBanner() {
        self.requestBanner()
    }

    @IBAction func goBack() {
        self.dismiss(animated: true, completion: nil)
    }

}
